frappe/erpnext
# fix: Clean Serial No input on Server Side

**Issue:**
- While entering serial nos often users add spaces at the end of the line or otherwise
- After https://github.com/frappe/erpnext/pull/24949, client side code to trim spaces was removed due to UX issues
- This allowed serial no string with random spaces to enter the db, and this failed the `get_stock_ledger_entries` query in `serial_no.py` 
  ![Screenshot 2021-07-22 at 12 18 48 PM](https://user-images.githubusercontent.com/25857446/126606533-6facc9ad-ffd9-4836-a5f5-20d00be8086f.png)
- As a result the serial nos with trailing spaces had **no Purchase Document attached against them**, because no stock ledger entry was found matching this serial no via the query.
- Partial Serial Nos had creation/purchase document updated against them

**Fix:**
- Clean serial nos on server side before inserting in db
- This is useful for faulty formats sent via APIs too, we shouldn't rely on client side code for such sanitation
- Since string is cleaned, query works fine

> I did think of altering the query so that it just searches for the serial no like `%123%`. But this may also get in highly inaccurate matches such as a string where 123 is a substring like `abc123def` and we don't want that.
The query searches for the following, making sure we match an **absolute whole serial no** :
>- serial no = '123'
>- serial no like '123\n'
>- serial no like '\n123'
>- serial no like '\n123\n'

**To test:**
- Create item with **Has Serial No** enabled, do not add Serial No Series
- Create a Purchase Receipt , for this item
- Add serial nos manually in line item with spaces at the end of the line like "ABC<space> DEF<space>"
- save submit
- Check if both serial nos are created and purchase documents are updated in both
- In Purchase receipt doc, in browser console, `cur_frm.doc.items[0].serial`, the value should be `ABC\nDEF` without spaces in between
